op-guide/binary-deployment.md

```diff
@@ -0,0 +1,486 @@
+| TiKV|  20160 |  TCP | the TiKV communication port  |
+| PD | 2379 | TCP | the communication port between TiDB and PD |
+| PD | 2380 | TCP | the inter-node communication port within the PD cluster |
+
+### TiDB database components (optional)
+
+See the following table for the default ports for the optional TiDB components:
+
+| Component | Default Port | Protocol | Description |
+| :-- | :-- | :-- | :------------------------ |
+| Prometheus |  9090| TCP | the communication port for the Prometheus service |
+| Pushgateway |  9091 | TCP | the aggregation and report port for TiDB, TiKV, and PD monitor |
+| Node_exporter|  9100| TCP | the communication port to report the system information of every TiDB cluster node |
+| Grafana | 3000 | TCP | the port for the external Web monitoring service and client (Browser) access |
+| alertmanager | 9093 | TCP | the port for the alert service |
+
+## Configure and check the system before installation
+
+### Operating system
+
+| Configuration | Description |
+| :-- | :-------------------- |
+| Supported Platform | See the [Software and Hardware Requirements](./recommendation.md) |
+| File System  |  The ext4 file system is recommended in TiDB Deployment |
+| Swap Space  |  The Swap Space is recommended to close in TiDB Deployment  |
+| Disk Block Size  |  Set the size of the system disk `Block` to `4096` |
+
+### Network and firewall
+
+| Configuration | Description |
+| :-- | :------------------- |
+| Firewall / Port | Check whether the ports required by TiDB are accessible between the nodes |
+
+### Operating system parameters
+
+| Configuration | Description |
+| :-- | :-------------------------- |
+| Nice Limits | For system users, set the default value of `nice` in TiDB to `0` |
+| min_free_kbytes | The setting for `vm.min_free_kbytes` in `sysctl.conf` needs to be high enough |
+| User Open Files Limit | For database administrators, set the number of TiDB open files to `1000000` |
+| System Open File Limits | Set the number of system open files to `1000000` |
+| User Process Limits | For TiDB users, set the `nproc` value to `4096` in `limits.conf` |
+| Address Space Limits | For TiDB users, set the space to `unlimited` in `limits.conf` |
+| File Size Limits | For TiDB users, set the `fsize` value to `unlimited` in `limits.conf` |
+| Disk Readahead | Set the value of the `readahead` data disk to `4096` at a minimum |
+| NTP service | Configure the NTP time synchronization service for each node |
+| SELinux  | Turn off the SELinux service for each node |
+| CPU Frequency Scaling | It is recommended to turn on CPU overclocking |
+| Transparent Hugepages | For Red Hat 7+ and CentOS 7+ systems, it is required to set the Transparent Hugepages to `always` |
+| I/O Scheduler | Set the I/O Scheduler of data disks to the `deadline` mode |
+| vm.swappiness | Set `vm.swappiness = 0` |
+
+
+> **Note**: To adjust the operating system parameters, contact your system administrator.
```

Comment:
I think we should be more specific than this.  As I understand, the only required change is to file limits.  We should show an example of how to change it.

Reply:
> use `sudo echo 100000 > /proc/sys/fs/file-max` to change User Open Files Limit to 100000
like this?

Reply:
@morgo Please see Jason's reply. If this is not what you mean, is it possible that you help add an example?

Reply:
@zz-jason afaik the sudo only applies to the echo there.  We could denote it is a root command with '#'

Reply:
@morgo @zz-jason The `echo` approach only applies to current session. When the system get rebooted, this setting will get lost. To make the setting persistent, users have to modify `/etc/security/limits.conf` like what tidb-ansible does https://github.com/pingcap/tidb-ansible/blob/master/roles/bootstrap/tasks/root_tasks.yml#L23-L31
```
      {{ deploy_user }}        soft        nofile        1000000
      {{ deploy_user }}        hard        nofile        1000000
      {{ deploy_user }}        soft        stack         10240
```
The `{{ deploy_user }}` is the user which runs tidb.

Reply:
@tennix good point.  How about:
```
export TIDBUSER="tidb" # change as required

cat << EOF > /tmp/tidb.conf
$TIDBUSER        soft        nofile        1000000
$TIDBUSER        hard        nofile        1000000
$TIDBUSER        soft        core          unlimited
$TIDBUSER        soft        stack         10240
EOF

sudo cp /tmp/tidb.conf /etc/security/limits.d/
sudo sysctl -w net.core.somaxconn=32768
sudo sysctl -w vm.swappiness=0
sudo sysctl -w net.ipv4.tcp_syncookies=0
sudo sysctl -w fs.file-max=1000000
```

This is what I used on my binary deployment [here](http://www.tocker.ca/this-blog-now-powered-by-wordpress-tidb.html).

Reply:
@morgo Note `sysctl -w` is similiar to `echo xx > /proc/xx` which only applies to current session. To persist the configurations, we also have to set this to system configuration file.
```shell
cat << EOF > /tmp/sysctl-tidb.conf
net.core.somaxconn = 32768
vm.swappiness = 0
net.ipv4.tcp_syncookies = 0
fs.file-max = 1000000
EOF

sudo cp /tmp/sysctl-tidb.conf /etc/sysctl.d/tidb.conf
sudo sysctl --system
```
I'm not sure how other kernel parameters are set, so just list the parameters you mentioned above.

Reply:
@tennix the above was designed to do both (I used `sysctl -w` to avoid the echo/sudo problem I mentioned above, and writing to a `/tmp` file and copy to apply on restart).  I am not familiar with `sysctl --system`, but if that triggers a reload, that's perfect.

Reply:
@zhangjinpeng1987 Could you help to explain which kernel parameters must be set and which are optional? Besides some parameters are inaccurate https://github.com/pingcap/docs/blob/0ad1c8e9183a19276cb7e9bd0b1c60c420db6764/op-guide/binary-deployment.md#operating-system-parameters for example `vm.min_free_kbytes` is suggested to be high enough, but what is high enough?

Reply:
@tennix @zhangjinpeng1987 I asked @LinuxGit the kernel parameters which already set in ansible are required, others are optional. 

BTW RHEL has a [PERFORMANCE TUNING GUIDE](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/performance_tuning_guide/), which explains the meaning of the kernel parameters and give some suggestions for adjustment. I think without a determined bottleneck, keep them as default maybe the best choice.

Reply:
I have a fair amount of confidence that only the `fs.file-max` limit needs to be changed. Context: https://github.com/tikv/tikv/issues/3387

@ethercflow I like the idea of leaving the others at default and linking to the guide with some suggestions of "what to look at".  @tennix how do you feel about only suggesting `fs.file-max` ?

Reply:
TiKV checks some other kernel parameters and reports warnings in the logs. I think it's better to tune these parameters to make TiKV happy. And I believe that's what TiDB Ansible does.

Reply:
This document does not actually involve parameter tuning. I think these parameters can be added, some parameters are consistent with `tidb-ansible`, some parameters are system defaults. Setting these parameters is good for performance, and no setting has no effect on deployment.

Reply:
For further improvement, we can do it later. As long as it does not affect the deployment and testing, I hope this PR can be merged soon. :)

Reply:
> and no setting has no affect on deployment.
@liubo0127 if they don't change `fs.file-max` it is possible tikv won't start in a really weird way. Context: tikv/tikv#3387

I am happy with it for now, but I would like to see them described as optional (except `fs.file-max`).